### ISSUE TITLE
Bug in `PseudoAttribute` equality in Python 3

### ISSUE THREAD

**Issue by dwhswenson** [openpathsampling/openpathsampling#721]
There's an error in the way `PseudoAttribute` tries to do equality in Python 3. Specifically, the approach uses attribute names that changed in Python 3. This is causing problems in #712.

Try the following complete test file:

```python
import os
import openpathsampling as paths
from nose.tools import assert_equal

import logging
logging.getLogger('openpathsampling.initialization').setLevel(logging.CRITICAL)
logging.getLogger('openpathsampling.ensemble').setLevel(logging.CRITICAL)
logging.getLogger('openpathsampling.storage').setLevel(logging.CRITICAL)
logging.getLogger('openpathsampling.netcdfplus').setLevel(logging.CRITICAL)

def test_problem():
    cv = paths.FunctionCV(name="x", f=lambda s: s.xyz[0][0])
    max_cv_func = lambda x, cv_: max(cv_(x))
    max_cv = paths.netcdfplus.FunctionPseudoAttribute(
        name="max x",
        key_class=paths.Trajectory,
        f=max_cv_func,
        cv_=cv
    )
    storage = paths.Storage("test_file.nc", 'w')
    storage.save(max_cv)
    storage.close()

    read_storage = paths.Storage("test_file.nc", 'r')
    read_max_cv = storage.attributes['max x']

    assert_equal(max_cv, read_max_cv)

    read_storage.close()
    if os.path.isfile("test_file.nc"):
        os.remove("test_file.nc")
```

Runs fine in Python 2. Fails in Python 3 (3.6, although based on errors in Travis, will get the same problem in 3.5) with:

```pytb
Traceback (most recent call last):
  File "/Users/dwhs/miniconda3/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/dwhs/Dropbox/msm-tis/openpathsampling/tests/test_problem.py", line 27, in test_problem
    assert_equal(max_cv, read_max_cv)
  File "/Users/dwhs/miniconda3/lib/python3.6/unittest/case.py", line 821, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/Users/dwhs/miniconda3/lib/python3.6/unittest/case.py", line 811, in _baseAssertEqual
    if not first == second:
  File "/Users/dwhs/Dropbox/msm-tis/openpathsampling/netcdfplus/attribute.py", line 331, in __eq__
    if hasattr(self.cv_callable.func_code, 'op_code') \
AttributeError: 'function' object has no attribute 'func_code'
```

Didn't we have equality check based on name of the CV? Isn't that why names are required for CVs? It looks like you're looking for the `op_code`, which, IIRC, was renamed in Py3k.

**Comment by jhprinz** [comment]
I think we used equality using a byte-code comparison, so in same python version the same code will be equal. The name is not really necessary, it used to be, when you wanted to save the cache, but I think even that is not really the case anymore. If I remember correctly, then I still use it in the cache name inside netcdf to make it easier to find the table.

Hmmm, either we 

1. use UUIDs as we do for most other objects, or
2. compare using the pickling algorithm we recently fixed
3. compare by name and kwargs (as we do now, but without the bytecode)

The fix is easy...

**Comment by dwhswenson** [comment]
> I think we used equality using a byte-code comparison, so in same python version the same code will be equal.

Just to be clear: the problem here isn't with the algorithm; it's with code that isn't compatible with Python 3. This isn't an assertion that isn't passing; the code will actually crash if you try to save a `FunctionPseudoAttribute` while running Python 3!

But it is worth considering the algorithm too:

> Hmmm, either we....

I don't think that it's totally obvious which approach is best. So here are my thoughts.

First, I think it is really important to distinguish between three types of objects:

* data (snapshots, trajectories, move changes, etc.)
* "how you are sampling" (path movers, etc.)
* "what you are sampling" (CVs, ensembles, volumes, etc.) 

Data object equality should always be UUID, because there should only ever be one instance of a snapshot with a specific set of values for all its features. That's easy.

"How" objects, are, I think, checked with UUID as well, although I think this is as much a matter of an inability to create good alternative check, short of overriding `__eq__` in every class (which may eventually be worth doing).

For several other "what" objects, such as volumes and ensembles, equality is determined based on creating a unique string describing that object. In this case, let's call the equality test on of "equivalence" (where the UUID-based approach is one of "identity"). This should make it easier to work with multiple simulations based on the same system, even if you recreate these objects (instead of reloading them from storage). This, by the way, is something else we might want to continue to make easier (in particular, by making sure that `__hash__` and `__eq__` use the same criteria.)

The thing that makes CVs different from ensembles/volumes is that the dictionaries associated with them mean that you can have difference (performance) behavior from two "equivalent" instances (e.g., if one has diskcache on and the other doesn't). That said, I would still lean toward considering "equivalence" as the basis of equality (which I think can be done with the pickling approach? is that correct?) and try to avoid duplication at the storage level. We already have a test to prevent multiple objects with the same name; is it possible to raise a warning if a user tries to save an CV with a different name, but the same bytecode/kwargs?

**Comment by dwhswenson** [comment]
Also, it sounds like that name is not as required for CVs as it once was. Anything we can do to get rid of the requirement that CVs have a name would be good. When teaching how to use OPS, that part is always a bit awkward. "Many objects in OPS can have names, but CVs *must* have names. And other objects, the name is added with `.named()`, but CVs, it's the first argument for initialization. Why do CVs require names? Storage."

**Comment by dwhswenson** [comment]
Another related thing to fix at the same time. Error message is misleading:

```pytb
  File "/Users/dwhs/Dropbox/msm-tis/openpathsampling/netcdfplus/stores/attribute.py", line 131, in add_diskcache
    'Need either at least one stored snapshot or a '
RuntimeError: Need either at least one stored snapshot or a template snapshot to determine type and shape of the CV.
```

Specifies "stored snapshot or a template snapshot," but what is needed is a template of whatever the `key_class` is (not always snapshot!)

**Comment by dwhswenson** [comment]
@jhprinz : Any progress on this? It's been a month, and this is blocking #712.

**Comment by dwhswenson** [comment]
> Data object equality should always be UUID, because there should only ever be one instance of a snapshot with a specific set of values for all its features. That's easy.

Thinking about this more -- this is still reasonable, but we would need to create conveniences in testing that check whether the fields of a snapshot are equal and whether the items of a trajectory are equal.

**Comment by jhprinz** [comment]
Okay, I will fix this tomorrow (cannot get my computer right now which runs pycharmm). But let me quickly summarize, that we want to keep the attempt to identify equivalence for CVs using the pickling approach. Basically just fix the code so it works in 3.6 and not alter the equivalence `__eq__` behavior. 

Sounds good. I will use the fixed code pickling and that should do it. Of course if you implement the same function with different code then this will still fail. But that is another topic.

About the name. I will think of this, but I believe this is the case. It used to be needed for the variable naming of the cache and I think to be human readable in the netCDF file it still is used, but could be replaced by the UUID of the object.
